mlrun/serving/server.py

```diff
@@ -526,6 +530,96 @@ def add_system_steps_to_graph(
     return graph
 
 
+async def _flush_all_batching_steps_single_pass(flow, logger, visited=None):
+    """
+    Flush all batching steps in a single pass through the graph.
+
+    Returns the total number of events flushed.
+    """
+    if visited is None:
+        visited = set()
+
+    events_flushed = 0
+
+    try:
+        # Walk through all steps in the flow
+        for step in flow._outlets:
+            # Skip if already visited (handles cyclic graphs)
+            if id(step) in visited:
+                continue
```

Comment:
As I understand you are doing multiple iterations saving `visited` across iterations, and then skipping an event on step revisit... But why? The whole point is to allow cyclic flows down the road, where you do have revisits..